cmd/clusterctl/pkg/client/repository/repository_github.go

```diff
@@ -0,0 +1,287 @@
+	"sigs.k8s.io/cluster-api/cmd/clusterctl/pkg/client/config"
+)
+
+const (
+	httpsScheme              = "https"
+	githubDomain             = "github.com"
+	gitHubTokeVariable       = "github-token"
+	gitHubReleaseRepository  = "releases"
+	gitHubLatestReleaseLabel = "latest"
+)
+
+// gitHubRepository provide support for providers hosted on GitHub.
+//
+// We support for GitHub repository that reading from GitHub releases.
+// Repository has full support for version, including also "latest" meta version.
+type gitHubRepository struct {
+	providerConfig           config.Provider
+	configVariablesClient    config.VariablesClient
+	authenticatingHTTPClient *http.Client
+	owner                    string
+	repository               string //this is the GitHub repository
+	defaultVersion           string
+	rootPath                 string
+	componentsPath           string
+	injectClient             *github.Client
+}
+
+var _ Repository = &gitHubRepository{}
+
+func (g *gitHubRepository) DefaultVersion() string {
+	return g.defaultVersion
+}
+
+func (g *gitHubRepository) RootPath() string {
+	return g.rootPath
+}
+
+func (g *gitHubRepository) ComponentsPath() string {
+	return g.componentsPath
+}
+
+func (g *gitHubRepository) GetFile(version, path string) ([]byte, error) {
+	release, err := g.getReleaseByTag(version)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get GitHub release %s", version)
+	}
+
+	// download files from the release
+	files, err := g.downloadFilesFromRelease(release, path)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to download files from GitHub release %s", version)
+	}
+
+	return files, nil
+
+}
+
+func newGitHubRepository(providerConfig config.Provider, configVariablesClient config.VariablesClient) (*gitHubRepository, error) {
+	if configVariablesClient == nil {
+		return nil, errors.New("invalid arguments: configVariablesClient can't be nil")
+	}
+
+	rURL, err := url.Parse(providerConfig.URL())
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid url")
+	}
+
+	// if the url is a github repository
+	if !(rURL.Scheme == httpsScheme && rURL.Host == githubDomain) {
+		return nil, errors.New("invalid url: a GitHub repository url should start with https://github.com")
+	}
+
+	// Checks if the repository Path is in the expected format:
+	t := strings.Split(strings.TrimPrefix(rURL.Path, "/"), "/")
+	if len(t) < 5 || t[2] != gitHubReleaseRepository {
+		return nil, errors.Errorf(
+			"invalid url: a GitHub repository url should be in the form https://github.com/{owner}/{Path}/%s/{latest|version-tag}/{componentsClient.yaml}",
+			gitHubReleaseRepository,
+		)
+	}
+
+	// extracts all the info from path
+	owner := t[0]
+	repository := t[1]
+	defaultVersion := t[3]
+	path := strings.Join(t[4:], "/")
+
+	// use path's directory as a rootPath
+	rootPath := filepath.Dir(path)
+	// use the file name (if any) as componentsPath
+	componentsPath := strings.TrimPrefix(strings.TrimPrefix(path, rootPath), "/")
```

Comment:
This is a little hard to follow, I'd suggest to make it a method and have some unit tests separately for it

Reply:
Added a new method `getComponentsPath`, tests will do in a follow PR